In [146]:
import os
os.getcwd()

'/Users/ashutoshsaini/eniac-discount-analysis/notebooks'

In [147]:
import pandas as pd
orders_df = pd.read_csv('../data/raw/orders.csv')

In [148]:
orders_orig = orders_df.copy()

In [149]:
orderlines_df = pd.read_csv('../data/raw/orderlines.csv')

In [150]:
products_df = pd.read_csv('../data/raw/products.csv')

In [151]:
brands_df = pd.read_csv('../data/raw/brands.csv')

In [152]:

orders_df.head(10)

order_id         created_date  total_paid      state
0    241319  2017-01-02 13:35:40       44.99  Cancelled
1    241423  2017-11-06 13:10:02      136.15  Completed
2    242832  2017-12-31 17:40:03       15.76  Completed
3    243330  2017-02-16 10:59:38       84.98  Completed
4    243784  2017-11-24 13:35:19      157.86  Cancelled
5    245275  2017-06-28 11:35:37      149.00  Completed
6    245595  2017-01-21 12:52:47      112.97  Completed
7    245851  2017-04-04 20:58:21       79.99    Pending
8    245941  2017-01-01 10:32:23      183.52  Completed
9    246018  2017-02-13 23:02:57      211.95  Completed

In [153]:
orders_df.nunique()

order_id        226909
created_date    224828
total_paid       31236
state                5
dtype: int64

In [154]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


In [155]:
orders_df.isnull().any()

order_id        False
created_date    False
total_paid       True
state           False
dtype: bool

In [156]:
orders_df.isnull().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

In [157]:
orders_df['order_id'].unique()

array([241319, 241423, 242832, ..., 527399, 527400, 527401])

In [158]:
state_mask = orders_df['state'] == 'Completed'
#state_mask
orders_df.loc[state_mask].nunique()

order_id        46605
created_date    46506
total_paid      12834
state               1
dtype: int64

In [159]:
#state_mask.isnull().sum 

#this is not helping me, my fault here was that I was using boolean on a Series and not a df.

In [160]:
completed_orders = orders_df.loc[state_mask]

completed_orders['total_paid'].isnull().sum()


np.int64(0)

In [161]:
#I would like to know the number of completed orders

state_mask.value_counts()

state
False    180304
True      46605
Name: count, dtype: int64

In [162]:
# I would now like to know the percentage of completed orders, just to make proportion of completed orders more visible

completed_orders_percentage = (state_mask.sum()/len(state_mask))*100

completed_orders_percentage

np.float64(20.53907072879436)

Only 20.54% of orders(completed orders) contribute to the total revenue of Eniac 

In [163]:
#I will now calculate the percenatge of pending and cancelled orders

pending_order_mask = orders_df['state'] == 'Pending'

pending_orders_count = (orders_df['state'] == 'Pending').sum()
percentage_of_pending_orders = (pending_orders_count/len(pending_order_mask))*100

percentage_of_pending_orders

np.float64(6.336901577284286)

In [164]:
cancelled_order_mask = orders_df['state'] == 'Cancelled'

cancelled_orders_count = (orders_df['state'] == 'Cancelled').sum()
percentage_of_cancelled_orders = (cancelled_orders_count/len(cancelled_order_mask))*100

percentage_of_cancelled_orders

np.float64(3.187621469399627)

# Now I will calculate the percentage share of Shopping Basket orders and Place Order state.

In [165]:
shopping_basket_order_mask = orders_df['state'] == 'Shopping Basket'

shopping_basket_order_mask_count = (orders_df['state'] == 'Shopping Basket').sum()
percentage_of_shopping_basket_order_mask = (shopping_basket_order_mask_count/len(shopping_basket_order_mask))*100

percentage_of_shopping_basket_order_mask

np.float64(51.9190512496199)

In [166]:
place_order_mask = orders_df['state'] == 'Place Order'

place_order_mask_count = (orders_df['state'] == 'Place Order').sum()
percentage_of_place_order_mask = (place_order_mask_count/len(place_order_mask))*100

percentage_of_place_order_mask

np.float64(18.017354974901835)

In [167]:
(orders_df['state'].value_counts(normalize=True) * 100)#.sum()

state
Shopping Basket    51.919051
Completed          20.539071
Place Order        18.017355
Pending             6.336902
Cancelled           3.187621
Name: proportion, dtype: float64

In [168]:
# sanity Check

total_state_masks_count = percentage_of_pending_orders + percentage_of_shopping_basket_order_mask + percentage_of_cancelled_orders + completed_orders_percentage + percentage_of_place_order_mask


total_state_masks_count

np.float64(100.00000000000001)

The Shopping Basket orders account for ~51.92% and Completed orders account for ~20.54%.

In [169]:
completed_orders = orders_df.loc[state_mask]

completed_orders_df = completed_orders.copy()

In [170]:
orderlines_df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [171]:
products_df.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

My next dataframe to explore is orderlines_df

In [172]:
orderlines_df_original = orderlines_df.copy()


In [ ]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [177]:
orderlines_df.head(15)

id  id_order  product_id  product_quantity      sku unit_price  \
0   1119109    299539           0                 1  OTT0133      18.99   
1   1119110    299540           0                 1  LGE0043     399.00   
2   1119111    299541           0                 1  PAR0071     474.05   
3   1119112    299542           0                 1  WDT0315      68.39   
4   1119113    299543           0                 1  JBL0104      23.74   
5   1119114    295310           0                10  WDT0249     231.79   
6   1119115    299544           0                 1  APP1582   1.137.99   
7   1119116    299545           0                 1  OWC0100      47.49   
8   1119119    299546           0                 1  IOT0014      18.99   
9   1119120    295347           0                 1  APP0700      72.19   
10  1119125    299548           0                 5  SPE0132      35.14   
11  1119126    299549           0                 1  PAC0929   2.565.99   
12  1119127    299550           0                 1  WDT0309     149.14   
13  1119129    299552           0                 1  MBI0010      16.99   
14  1119130    299552           0                 1  SAT0044      19.99   

                   date  
0   2017-01-01 00:07:19  
1   2017-01-01 00:19:45  
2   2017-01-01 00:20:57  
3   2017-01-01 00:51:40  
4   2017-01-01 01:06:38  
5   2017-01-01 01:14:27  
6   2017-01-01 01:17:21  
7   2017-01-01 01:46:16  
8   2017-01-01 01:50:34  
9   2017-01-01 01:54:11  
10  2017-01-01 02:02:20  
11  2017-01-01 02:07:42  
12  2017-01-01 02:09:52  
13  2017-01-01 02:13:51  
14  2017-01-01 02:14:36

In [178]:
orderlines_df.isnull().sum()

id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64

In [180]:
orderlines_df.nunique()

id                  293983
id_order            204855
product_id               1
product_quantity        67
sku                   7951
unit_price           11329
date                251631
dtype: int64

In [251]:
orderlines_df.duplicated().sum()

np.int64(0)

I will now change the Datatypes for date column to Datetime and unit_price to int

In [194]:
orderlines_df['date'] = pd.to_datetime(orderlines_df['date'], errors = 'coerce')

In [ ]:
orderlines_df['unit_price'] = pd.to_numeric(orderlines_df['unit_price'], errors = 'coerce')



In [195]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        257814 non-null  float64       
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 15.7+ MB


In [200]:
orderlines_df.isnull().sum()



id                      0
id_order                0
product_id              0
product_quantity        0
sku                     0
unit_price          36169
date                    0
dtype: int64

In [202]:
orderlines_df.loc[:,['unit_price','sku']].isnull().value_counts()

unit_price  sku  
False       False    257814
True        False     36169
Name: count, dtype: int64

Finally after exploring the orderlines_df I will move to products_df

In [ ]:
products_df_orig = products_df.copy() 

In [204]:
products_df.head(20)

sku                                               name  \
0     RAI0007                  Silver Rain Design mStand Support   
1     APP0023                  Apple Mac Keyboard Keypad Spanish   
2     APP0025                   Mighty Mouse Apple Mouse for Mac   
3     APP0072      Apple Dock to USB Cable iPhone and iPod white   
4     KIN0007        Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
5     APP0073     Apple Composite AV Cable iPhone and iPod white   
6     KIN0008        Mac Memory Kingston 1GB 667MHz DDR2 SO-DIMM   
7     KIN0009        Mac Memory Kingston 2GB 800MHz DDR2 SO-DIMM   
8   KIN0001-2  Mac memory Kingston 4GB (2x2GB) 667MHz DDR2 SO...   
9     APP0100             Apple Adapter Mini Display Port to VGA   
10    APP0101              Apple Mini DisplayPort to DVI Adapter   
11    SEN0021  Sennheiser CX 300-II Precision headphones iPho...   
12    APP0111                     Apple USB Ethernet Adapter Mac   
13    APP0156              45W MagSafe charger Apple MacBook Air   
14    MOS0013         Adapter Moshi FireWire 400 to FireWire 800   
15    MOS0021               Clearguard Moshi MacBook Pro and Air   
16    APP0206  60W MagSafe charger Apple MacBook and MacBook ...   
17    HGD0007       Henge Docks Docking Station MacBook Pro 13 "   
18    MUV0009    Muvit Back Clear Case iPhone 4 / 4S Transparent   
19    APP0234                        Apple Dock Connector to VGA   

                                                 desc  price promo_price  \
0        Aluminum support compatible with all MacBook  59.99     499.899   
1           USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                              mouse Apple USB cable.     59     569.898   
3               IPhone dock and USB Cable Apple iPod.     25     229.997   
4   2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
5   IPhone and iPod AV Cable Dock to Composite Video.     45     420.003   
6   1GB RAM Mac mini and iMac (2006/07) MacBook Pr...  18.99     146.471   
7        2GB RAM iMac with Intel Core 2 Duo (Penryn).  36.99     274.694   
8   RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...     74     669.904   
9   Adapter Mini Display Port to VGA MacBook and M...     35     330.003   
10  Adapter mini DisplayPort to DVI MacBook and Ma...     35     330.003   
11     Headphones iPhone iPad iPad 2 iPad 3 and iPod.  49.99     449.878   
12                       USB to Ethernet adapter Mac.     35     330.003   
13     Charger MacBook Air MagSafe 45 W (MC747Z / A).     89     849.904   
14           FireWire 400 adapter cable FireWire 800.     20     179.891   
15  Keyboard Protector MacBook Pro 13-inch Retina ...  24.95     249.889   
16  Charger MacBook and 13 inch MacBook MagSafe 60...     89     849.904   
17  Dock and support with multiple connections for...  69.99     699.864   
18                  Case iPhone 4 / 4S polycarbonate.   9.99       7.986   
19                         Dock Connector to VGA IOS.     35     350.005   

    in_stock      type  
0          1      8696  
1          0  13855401  
2          0      1387  
3          0      1230  
4          1      1364  
5          0      1230  
6          0      1364  
7          0      1364  
8          0      1364  
9          0      1325  
10         0      1325  
11         0      5384  
12         0      1334  
13         0  13005399  
14         0      1325  
15         0  13835403  
16         1  13005399  
17         0  12995397  
18         0  11865403  
19         0  13955395

In [205]:
products_df.isnull().sum()

sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64

In [214]:
products_df.duplicated().value_counts()

False    10580
True      8746
Name: count, dtype: int64

In [226]:
duplicated_rows_mask = products_df['sku'].duplicated()

products_df.loc[duplicated_rows_mask]

sku                                               name  \
101    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
102    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
103    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
104    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
105    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
...        ...                                                ...   
16831  APP2302  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
16833  APP2303  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
18190  PAR0077                         Parrot Bebop Drone 2 Power   
18308  NKI0010              Nokia Wireless sphygmomanometer Plata   
18514  PLA0035          Bluetooth Headset Plantronics Explorer 80   

                                                    desc        price  \
101          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
102          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
103          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
104          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
105          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
...                                                  ...          ...   
16831  New MacBook Pro 13-inch Core i5 Touch Bar 33 G...   26.155.941   
16833  New MacBook Pro 13 inch Touch Bar 33 GHz Core ...  237.559.421   
18190  Drone cuadricóptero quality camera integrated ...        699.9   
18308     Sphygmomanometer for iPhone iPad and iPod App.       129.99   
18514  egonómico wireless headset design for iPhone i...        39.99   

      promo_price  in_stock      type  
101    11.455.917         0      1282  
102    11.455.917         0      1282  
103    11.455.917         0      1282  
104    11.455.917         0      1282  
105    11.455.917         0      1282  
...           ...       ...       ...  
16831  26.155.941         0  1,02E+12  
16833  23.755.942         0  1,02E+12  
18190   6.733.892         0  11905404  
18308   1.149.899         1  11905404  
18514     349.896         1      5384  

[8747 rows x 7 columns]

In [230]:
duplicated_sku_mask = products_df['sku'] == 'APP0390'

In [237]:
duplicate_sku = products_df.loc[duplicated_sku_mask]

duplicate_sku.duplicated().sum()


np.int64(8)

In [235]:
duplicate_sku.nunique()

sku            1
name           1
desc           1
price          1
promo_price    1
in_stock       1
type           1
dtype: int64

In [240]:
duplicated_skus = products_df[products_df['sku'].duplicated(keep=False)]
duplicated_skus

sku                                               name  \
100    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
101    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
102    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
103    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
104    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
...        ...                                                ...   
18190  PAR0077                         Parrot Bebop Drone 2 Power   
18307  NKI0010              Nokia Wireless sphygmomanometer Plata   
18308  NKI0010              Nokia Wireless sphygmomanometer Plata   
18513  PLA0035          Bluetooth Headset Plantronics Explorer 80   
18514  PLA0035          Bluetooth Headset Plantronics Explorer 80   

                                                    desc   price promo_price  \
100          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
101          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
102          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
103          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
104          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
...                                                  ...     ...         ...   
18190  Drone cuadricóptero quality camera integrated ...   699.9   6.733.892   
18307     Sphygmomanometer for iPhone iPad and iPod App.  129.99   1.149.899   
18308     Sphygmomanometer for iPhone iPad and iPod App.  129.99   1.149.899   
18513  egonómico wireless headset design for iPhone i...   39.99     349.896   
18514  egonómico wireless headset design for iPhone i...   39.99     349.896   

       in_stock      type  
100           0      1282  
101           0      1282  
102           0      1282  
103           0      1282  
104           0      1282  
...         ...       ...  
18190         0  11905404  
18307         1  11905404  
18308         1  11905404  
18513         1      5384  
18514         1      5384  

[9504 rows x 7 columns]

Important part starts now, this includes the removal of the hazardous skus from the dataframe

In [242]:
duplicated_skus.groupby('sku').nunique()

name  desc  price  promo_price  in_stock  type
sku                                                      
AII0013       1     1      1            1         1     1
APP0390       1     1      1            1         1     1
APP0574       1     1      1            1         1     1
APP0624       1     1      1            1         1     1
APP0733       1     1      1            1         1     1
...         ...   ...    ...          ...       ...   ...
SYN0140       1     1      1            1         1     1
SYN0141       1     1      1            1         1     1
SYN0142       1     1      1            1         1     1
WDT0311-A     1     1      1            1         1     1
WDT0317       1     1      1            1         1     1

[757 rows x 6 columns]

In [ ]:
'''PLEASE DISCARD THIS CODE SNIPPET, 

WHY: BECAUSE IT DROPPED A LOT OF ROWS AND EVEN WITH ONE SKU CONSIDERING IT AS A DUPLICATE, WHIHC IS NOT THE RIGHT APPROACH. 

THEREFORE I DECIDED TO HAVE ONE ROW PER SKU AND NOT 0 ROWS PER DUPLICATED SKUs

FOR EX, IT DELETED ENTIRELY THAT SKU FROM THE DATAFRAME WHIHC WAS A DUPLICATE, MEANING IF THERE IS AN SKU AP033090 IT DELETED THAT COMPLETELY AND WE DONT WANT THAT

WE WANT THAT SKU BUT ONLY ONES PER ROW'''

#products_df_1 = products_df.drop_duplicates(subset='sku', keep = False)

#products_df_1


'PLEASE DISCARD THIS CODE SNIPPET, \n\nWHY: BECAUSE IT DROPPED A LOT OF ROWS AND EVEN WITH ONE SKU REPEATING. \n\nTHEREFORE I DECIDED TO HAVE ONE ROW PER SKU AND NOT 0 ROWS PER REPEATING SKUs'

The code snippet which you see right below, ... this cell has one trick, guess what, i used a parameter keep= 'first' this allows us to safeguard each row per unique skus which is much needed rather than just deleting all skus which are duplicated. So we have now no duplicate skus from the cell below.

In [247]:
products_df = products_df.drop_duplicates(subset = 'sku', keep = 'first')

products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19321         1     12282  
19322         1      1392  
19323         1      1392  
19324         0      1392  
19325         1      1392  

[10579 rows x 7 columns]

another step is to remove promo_price column. Why I want to remove promo_price, because it is not adding any useful value, instead creating noise in the dataframe, my golden source of truth to look for discounting strategy lies in the columns price in products_df adn unit_price in orderlines_df 

In [252]:
products_df.isnull().sum()

sku             0
name            0
desc            7
price          45
promo_price     0
in_stock        0
type           50
dtype: int64

In [255]:
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19321         1     12282  
19322         1      1392  
19323         1      1392  
19324         0      1392  
19325         1      1392  

[10579 rows x 7 columns]

In [254]:
orderlines_df.isnull().sum()

id                      0
id_order                0
product_id              0
product_quantity        0
sku                     0
unit_price          36169
date                    0
dtype: int64

In [261]:
merged_product_orderline_df = orderlines_df.merge(products_df, left_on='sku',right_on='sku',how='left')

merged_product_orderline_df

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
293978  1650199    527398           0                 1  JBL0122       42.99   
293979  1650200    527399           0                 1  PAC0653      141.58   
293980  1650201    527400           0                 2  APP0698        9.99   
293981  1650202    527388           0                 1  BEZ0204       19.99   
293982  1650203    527401           0                 1  APP0927       13.99   

                      date                                               name  \
0      2017-01-01 00:07:19  Otterbox iPhone Case Symmetry 2.0 SE / 5s / 5 ...   
1      2017-01-01 00:19:45         27UD58-B LG Monitor 27 "4K UHD DisplayPort   
2      2017-01-01 00:20:57  Parrot Bebop 2 White + Command FLYPAD and FPV ...   
3      2017-01-01 00:51:40              Blue WD 2TB Hard Drive 35 "Mac and PC   
4      2017-01-01 01:06:38                      Gray Bluetooth Speaker JBL GO   
...                    ...                                                ...   
293978 2018-03-14 13:57:25                JBL T450 BT Bluetooth Headset Black   
293979 2018-03-14 13:57:34  Samsung SSD 850 expansion kit EVO 250GB + Data...   
293980 2018-03-14 13:57:41  Apple Lightning Cable Connector to USB 1m Whit...   
293981 2018-03-14 13:58:01            Be.ez LArobe Case Mix Macbook 12 "Green   
293982 2018-03-14 13:58:36  EarPods Apple Headphones with Remote and Mic (...   

                                                     desc   price promo_price  \
0       resistant cover and thin beveled edges for iPh...   34.99     199.904   
1       Monitor for gamers and multimedia professional...     429   3.989.999   
2       cuadricóptero wireless remote control with 25 ...     699   5.689.989   
3       Internal Hard Drive Western Digital 2TB 3.5-in...      79     639.945   
4       Compact Bluetooth Handsfree for iPhone iPad an...    29.9       27.99   
...                                                   ...     ...         ...   
293978  Wireless headphones with folding design with 1...   49.95     429.901   
293979  SSD upgrade kit 2008-2010 250 GB MacBook and M...  215.98   1.415.845   
293980  Apple Lightning USB Cable 1 meter to charge an...      25      99.898   
293981                     Macbook thin sheath 12 inches.   29.99     199.904   
293982  EarPods headphones Apple iPhone iPad and iPod ...      35       13.99   

        in_stock      type  
0            0.0  11865403  
1            0.0      1296  
2            0.0  11905404  
3            0.0  12655397  
4            1.0      5398  
...          ...       ...  
293978       1.0      5384  
293979       1.0      1433  
293980       1.0      1230  
293981       0.0  13835403  
293982       1.0      5384  

[293983 rows x 13 columns]

In [265]:
#merged_product_orderline_df.drop(['promo_price', 'type'], axis=1, inplace=True)

merged_product_orderline_df

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
293978  1650199    527398           0                 1  JBL0122       42.99   
293979  1650200    527399           0                 1  PAC0653      141.58   
293980  1650201    527400           0                 2  APP0698        9.99   
293981  1650202    527388           0                 1  BEZ0204       19.99   
293982  1650203    527401           0                 1  APP0927       13.99   

                      date                                               name  \
0      2017-01-01 00:07:19  Otterbox iPhone Case Symmetry 2.0 SE / 5s / 5 ...   
1      2017-01-01 00:19:45         27UD58-B LG Monitor 27 "4K UHD DisplayPort   
2      2017-01-01 00:20:57  Parrot Bebop 2 White + Command FLYPAD and FPV ...   
3      2017-01-01 00:51:40              Blue WD 2TB Hard Drive 35 "Mac and PC   
4      2017-01-01 01:06:38                      Gray Bluetooth Speaker JBL GO   
...                    ...                                                ...   
293978 2018-03-14 13:57:25                JBL T450 BT Bluetooth Headset Black   
293979 2018-03-14 13:57:34  Samsung SSD 850 expansion kit EVO 250GB + Data...   
293980 2018-03-14 13:57:41  Apple Lightning Cable Connector to USB 1m Whit...   
293981 2018-03-14 13:58:01            Be.ez LArobe Case Mix Macbook 12 "Green   
293982 2018-03-14 13:58:36  EarPods Apple Headphones with Remote and Mic (...   

                                                     desc   price  in_stock  
0       resistant cover and thin beveled edges for iPh...   34.99       0.0  
1       Monitor for gamers and multimedia professional...     429       0.0  
2       cuadricóptero wireless remote control with 25 ...     699       0.0  
3       Internal Hard Drive Western Digital 2TB 3.5-in...      79       0.0  
4       Compact Bluetooth Handsfree for iPhone iPad an...    29.9       1.0  
...                                                   ...     ...       ...  
293978  Wireless headphones with folding design with 1...   49.95       1.0  
293979  SSD upgrade kit 2008-2010 250 GB MacBook and M...  215.98       1.0  
293980  Apple Lightning USB Cable 1 meter to charge an...      25       1.0  
293981                     Macbook thin sheath 12 inches.   29.99       0.0  
293982  EarPods headphones Apple iPhone iPad and iPod ...      35       1.0  

[293983 rows x 11 columns]